### PR TITLE
Cognito User Pool UI Customization

### LINKED ISSUE #3634: Support for Cognito user pool client UI customizations
Hi,

AWS supports [customizing Cognito UserPool app clients](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-app-ui-customization.html) with CSS and a logo.

image

This can be done in the AWS Console UI and using the CLI. It doesn't seem to be available in CloudFormation yet.

- https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/get-ui-customization.html
- https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/set-ui-customization.html

I don't think the Terraform provider currently supports this. Is this on the roadmap?

- Terraform Version = `0.11.3`
- Affected Resource(s)
  - [cognito_user_pool_client](https://www.terraform.io/docs/providers/aws/r/cognito_user_pool_client.html)

Thanks!

### PR DESCRIPTION
**Pull request by gdlx**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #3634

Changes proposed in this pull request:

* Ability to define Cognito User Pool Client UI Customization
* Ability to define Cognito User Pool UI Customization (for uncustomized clients, as managed by AWS API)

I've written acceptance tests but didn't run them (I've tested the feature in an existing workspace).

As the feature is usable from 2 different resources, I've tried to avoid code duplication as much as possible.

I've moved `loadFileContent` function from `resource_aws_lambda_function.go` to `utils.go`  to make it global.

The `remoteFileContent` function in `utils.go` is kind of a hack to allow getting content seamlessly from a local file or a URL, because we send a file to AWS API, but it sends back an URL.

### CODE REVIEW

File: website/docs/r/cognito_user_pool.markdown [hunk 1]
```diff
@@ -127,6 +128,11 @@ The following arguments are supported:
   * `email_subject_by_link` (Optional) - The subject line for the email message template for sending a confirmation link to the user.
   * `sms_message` (Optional) - The SMS message template. Must contain the `{####}` placeholder. Conflicts with `sms_verification_message` argument.
 
+#### UI Customization
+
+  * `css` (Optional) - The customized CSS.
+  * `image_file` (Optional) - Le local image file path.
```
Comment (chadasapp) [inline website/docs/r/cognito_user_pool.markdown]: Utilisez "the" ;-)

File: website/docs/r/cognito_user_pool_client.markdown [hunk 2]
```diff
@@ -60,6 +82,12 @@ The following arguments are supported:
 * `supported_identity_providers` - (Optional) List of provider names for the identity providers that are supported on this client.
 * `user_pool_id` - (Required) The user pool the client belongs to.
 * `write_attributes` - (Optional) List of user pool attributes the application client can write to.
+* `ui_customization` (Optional) - The [UI Customization](#ui-customization).
+
+#### UI Customization
+
+  * `css` (Optional) - The customized CSS.
+  * `image_file` (Optional) - Le local image file path.
```
Comment (chadasapp) [inline website/docs/r/cognito_user_pool_client.markdown]: Same.

File: aws/resource_aws_cognito_user_pool_client.go [hunk 3]
```diff
@@ -218,6 +220,10 @@ func resourceAwsCognitoUserPoolClientCreate(d *schema.ResourceData, meta interfa
 
 	d.SetId(*resp.UserPoolClient.ClientId)
 
+	if err = cognitoUserPoolUICustomizationSet(d, conn); err != nil {
+		return fmt.Errorf("Error setting User Pool Client UI Customization: %s", err)
```
Comment (chadasapp) [inline aws/resource_aws_cognito_user_pool_client.go]: Add "Cognito" before "User Pool Client" to be consistent with the error below.

File: aws/resource_aws_cognito_user_pool.go [hunk 4]
```diff
@@ -921,6 +937,10 @@ func resourceAwsCognitoUserPoolUpdate(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error updating Cognito User pool: %s", err)
 	}
 
+	if err = cognitoUserPoolUICustomizationSet(d, conn); err != nil {
+		return fmt.Errorf("Error updating User Pool UI Customization: %s", err)
```
Comment (chadasapp) [inline aws/resource_aws_cognito_user_pool.go]: Same thing: error consistency.

File: aws/resource_aws_cognito_user_pool.go [hunk 5]
```diff
@@ -660,6 +662,10 @@ func resourceAwsCognitoUserPoolCreate(d *schema.ResourceData, meta interface{})
 
 	d.SetId(*resp.UserPool.Id)
 
+	if err = cognitoUserPoolUICustomizationSet(d, conn); err != nil {
+		return fmt.Errorf("Error setting User Pool UI Customization: %s", err)
```
Comment (chadasapp) [inline aws/resource_aws_cognito_user_pool.go]: Same thing, error consistency.

**Comment by gdlx** [comment]
Thanks @chadasapp !

I'm currently in holiday so I'll unlikely be able to look at this before the end of august.

> Utilisez "the" ;-)

No idea why I've written this, I never use French in code 🤔

File: website/docs/r/cognito_user_pool.markdown (hunk 1 above)
Comment (timmjd) [reply website/docs/r/cognito_user_pool.markdown]:
```suggestion
  * `image_file` (Optional) - The local image file path.
```

File: website/docs/r/cognito_user_pool_client.markdown (hunk 2 above)
Comment (timmjd) [reply website/docs/r/cognito_user_pool_client.markdown]:
```suggestion
  * `image_file` (Optional) - The local image file path.
```

File: aws/resource_aws_cognito_user_pool_client.go (hunk 3 above)
Comment (timmjd) [reply aws/resource_aws_cognito_user_pool_client.go]:
```suggestion
		return fmt.Errorf("Error setting Cognito User Pool Client UI Customization: %s", err)
```

File: aws/resource_aws_cognito_user_pool.go (hunk 5 above)
Comment (timmjd) [reply aws/resource_aws_cognito_user_pool.go]:
```suggestion
		return fmt.Errorf("Error setting Cognito User Pool UI Customization: %s", err)
```

File: aws/resource_aws_cognito_user_pool.go (hunk 4 above)
Comment (timmjd) [reply aws/resource_aws_cognito_user_pool.go]:
```suggestion
		return fmt.Errorf("Error updating Cognito User Pool UI Customization: %s", err)
```

**Comment by gdlx** [comment]
> Hey @gdlx, are you back from vacation? Would be great to get this feature

Sorry, I didn't work with Cognito nor Terraform for a few months so it's a bit difficult to go back to this but I'll take a look shortly.

**Comment by gdlx** [comment]
@timmjd I've accepted your changes, thanks a lot for helping ! But I fear this won't be enough to get the feature merged. It has to be reviewed by the core team.

**Comment by renatoargh** [comment]
I think that CI needs to pass as well, @gdlx - this will be a very useful feature, hope someone with some terraform foo can help you get this merged 🥇

**Comment by mubeta06** [comment]
👍

File: aws/cognito_user_pool_ui_customization.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"css": {
+					Type:     schema.TypeString,
+					Optional: true,
+					DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+						return new == stringHashSum(old)
+					},
+					StateFunc: func(v interface{}) string {
+						switch v.(type) {
```
Comment (mubeta06) [inline aws/cognito_user_pool_ui_customization.go]: Looks as though some minor changes are required here to get the tests passing lints. `switch v := v.(type)` and remove the string type assertion.

File: aws/cognito_user_pool_ui_customization.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+					},
+				},
+				"image_file": {
+					Type:     schema.TypeString,
+					Optional: true,
+					DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+						return new == remoteFileHashSum(old)
+					},
+					StateFunc: func(v interface{}) string {
+						switch v.(type) {
```
Comment (mubeta06) [inline aws/cognito_user_pool_ui_customization.go]: And the same here.
`switch v:= v.(type)`

**Comment by gdlx** [comment]
I'm sorry, I don't work with Terraform anymore for now and don't have much time to improve the PR.
I mean, I could blindly apply your change proposals but I'd have no way to test the code behavior.

That and the fact that the tests were passing when I sent the PR and it didn't help merging it then, it doesn't motivate me to spend more time to wait 6 more months before having to fix it again.

**However, I'd be pleased to give push access on the repo to whoever wants to fix the code !**

**Comment by jfharden** [comment]
@gdlx I'll pick this up if you are ok with that?

**Comment by supergibbs** [comment]
@jfharden Did you get access? Happy to help test if you need it. I have a terraform script for Cognito but manually use CLI for this part.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by alejandrod** [comment]
This will be VERY useful.

**Comment by breathingdust** [comment]
Hi all! :wave: Just wanted to direct you to our public [roadmap](https://github.com/hashicorp/terraform-provider-aws/blob/master/ROADMAP.md) for this quarter (Nov-Jan) in which this item has been mentioned.

Due to the significant community interest in support for this feature, we will be looking at merging existing contributions soon. 

We appreciate all the contributions and feedback thus far.

Look out for support in the provider soon!

**Comment by alejandrod** [comment]
Waiting for this! Very important. I hope it gets merged soon.

**Comment by opub** [comment]
@breathingdust did this get bumped from last quarter's release scope? It doesn't look like there has been much progress on it and don't see it in the Feb-Apr road map. I'm not familiar with the Terraform release process though.

**Comment by anGie44** [comment]
Hi @gdlx 👋 Thank you so much for this PR! Just dropping a note that I've rebased and will tidy up any linter errors that may come up so the team can review the work you've contributed here.

**Comment by anGie44** [comment]
Output of acceptance tests (commercial):

```
--- PASS: TestAccAWSCognitoUserPoolClient_disappears (16.75s)
--- PASS: TestAccAWSCognitoUserPoolClient_allFields (21.51s)
--- PASS: TestAccAWSCognitoUserPoolClient_basic (23.06s)
--- PASS: TestAccAWSCognitoUserPoolClient_analyticsConfigWithArn (26.42s)
--- PASS: TestAccAWSCognitoUserPoolClient_RefreshTokenValidity (32.59s)
--- PASS: TestAccAWSCognitoUserPoolClient_allFieldsUpdatingOneField (34.37s)
--- PASS: TestAccAWSCognitoUserPoolClient_Name (37.37s)
--- PASS: TestAccAWSCognitoUserPoolClient_analyticsConfig (49.08s)
--- PASS: TestAccAWSCognitoUserPoolClient_UICustomization_ImageFile (56.67s)
--- PASS: TestAccAWSCognitoUserPoolClient_UICustomization_CSSandImageFile (57.25s)
--- PASS: TestAccAWSCognitoUserPoolClient_UICustomization_CSS (59.16s)
```
```
--- PASS: TestAccAWSCognitoUserPool_SmsConfiguration_ExternalId (564.96s)
--- PASS: TestAccAWSCognitoUserPool_withPasswordPolicy (566.11s)
--- PASS: TestAccAWSCognitoUserPool_SmsAuthenticationMessage (611.32s)
--- PASS: TestAccAWSCognitoUserPool_MfaConfiguration_SmsConfigurationAndSoftwareTokenMfaConfiguration (612.74s)
--- PASS: TestAccAWSCognitoUserPool_withEmailVerificationMessage (522.71s)
--- PASS: TestAccAWSCognitoUserPool_withAdminCreateUserConfigurationAndPasswordPolicy (254.85s)
--- PASS: TestAccAWSCognitoUserPool_withTags (693.87s)
--- PASS: TestAccAWSCognitoUserPool_withAdvancedSecurityMode (694.62s)
--- PASS: TestAccAWSCognitoUserPool_SmsConfiguration (722.14s)
--- PASS: TestAccAWSCognitoUserPool_withUICustomization_CSS (732.47s)
--- PASS: TestAccAWSCognitoUserPool_update (743.26s)
--- PASS: TestAccAWSCognitoUserPool_withAdminCreateUserConfiguration (438.97s)
--- PASS: TestAccAWSCognitoUserPool_withDeviceConfiguration (315.77s)
--- PASS: TestAccAWSCognitoUserPool_MfaConfiguration_SoftwareTokenMfaConfigurationToSmsConfiguration (288.17s)
--- PASS: TestAccAWSCognitoUserPool_withUICustomization_CSSandImageFile (810.03s)
--- PASS: TestAccAWSCognitoUserPool_MfaConfiguration_SmsConfigurationToSoftwareTokenMfaConfiguration (272.14s)
--- PASS: TestAccAWSCognitoUserPool_withUICustomization_ImageFile (824.50s)
--- PASS: TestAccAWSCognitoUserPool_MfaConfiguration_SmsConfiguration (513.44s)
--- PASS: TestAccAWSCognitoUserPool_MfaConfiguration_SoftwareTokenMfaConfiguration (343.25s)
--- PASS: TestAccAWSCognitoUserPool_recovery (348.39s)
```

**Comment by anGie44** [comment]
Hi @gdlx 👋 after more robust testing, I came across some unfriendly resource behaviors when configuring the UI Customization feature w/in the existing User Pool and User Pool Client resources including:
* the computed behavior of the configuration block when it's applied on a user pool and thus is returned on a state refresh of a user pool client
* possibility of conflicting settings when configured in both resources

With that said, i've moved the "ui_customization" block out into it's own managed resource. I believe there, it'll be more apparent where the customization is applied and updates/deletes can occur without affecting the state of the existing user pool and/or user pool clients.

**Pull request merged**